File: src/base/go2_twist_bridge_py/go2_twist_bridge_py/twist_bridge.py
#包含头文件
import rclpy
from rclpy.node import Node
from unitree_api.msg import Request
from geometry_msgs.msg import Twist
import json
from .sport_model import ROBOT_SPORT_API_IDS

# 自定义节点
class TwistBridge(Node):
    def __init__(self):
        super().__init__("twist_brige")
        self.get_logger().info("启动成功")
         # 1. 创建一个Request的发布对象
        self.request_pub = self.create_publisher(Request, '/api/sport/request', 10)

        # 2. 创建twist订阅对象
        self.twist_sub = self.create_subscription(
            Twist,
            'cmd_vel',
            self.twist_cb,
            10
        )

    def twist_cb(self, twist):
        # 3. 在回调函数中实现消息的转换以及发布
        request_msg = Request()

        # 获取 twist 消息的线速度和角速度
        x = twist.linear.x
        y = twist.linear.y
        z = twist.angular.z

        # 默认api_id 平衡站立
        api_id = ROBOT_SPORT_API_IDS["BALANCESTAND"]
        parameter = ''

        if x != 0 or y != 0 or z != 0:
            api_id = ROBOT_SPORT_API_IDS["MOVE"]
            # 设置参数 -- 字符串样式的速度指令
            js = {
                'x': x,
                'y': y,
                'z': z
            }
            parameter = json.dumps(js)

        request_msg.header.identity.api_id = api_id
        request_msg.parameter = parameter

        # 发布Request消息
        self.request_pub.publish(request_msg)
        self.get_logger().info('Published request message')


def main(args=None):
    rclpy.init(args=args)
    twist_bridge = TwistBridge()
    rclpy.spin(twist_bridge)
    twist_bridge.destroy_node()
    rclpy.shutdown()

if __name__=='__main__':
   main()

File: src/base/go2_twist_bridge/src/twist_bridge.cpp
/*
    需求：订阅twist消息，并将之转换为个go2所需的request消息以控制机器狗运动
    实现：
      1. 创建一个Request的发布对象
      2. 创建twist订阅对象
      3. 在回调函数中实现消息的转换以及发布。


*/

//包含头文件
#include "rclcpp/rclcpp.hpp"
#include "unitree_api/msg/request.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "sport_model.hpp"
#include "nlohmann/json.hpp"
using namespace std::placeholders;

// 自定义节点类
class TwistBridge :public rclcpp::Node
{
    public:
    TwistBridge():Node("My_node_node")
    {
        RCLCPP_INFO(this->get_logger(),"TwistBridge创建成功");
        // 1. 创建一个Request的发布对象
        request_pub_ = this->create_publisher<unitree_api::msg::Request>("/api/sport/request",10);
        twist_sub_ = this->create_subscription<geometry_msgs::msg::Twist>("cmd_vel",10,std::bind(&TwistBridge::twist_cb,this,_1));
        // 2. 创建twist订阅对象
    }
    private:
    rclcpp::Publisher<unitree_api::msg::Request>::SharedPtr request_pub_;
    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr twist_sub_;
    void twist_cb(const geometry_msgs::msg::Twist::SharedPtr twist){
        // 3. 在回调函数中实现消息的转换以及发布
        unitree_api::msg::Request request_msg;
        // 转换
        // 获取 twist 消息的线速度和角速度
        double x = twist->linear.x;
        double y = twist->linear.y;
        double z = twist->angular.z;
        // 默认api_id 平衡站立
        auto api_id = ROBOT_SPORT_API_ID_BALANCESTAND;
        if (x != 0 || y != 0 || z != 0)
        {
          api_id = ROBOT_SPORT_API_ID_MOVE;
          // 设置参数 -- 字符串样式的速度指令
          nlohmann::json js;
          js["x"] = x;
          js["y"] = y;
          js["z"] = z;
          request_msg.parameter = js.dump();
        }
        request_msg.header.identity.api_id = api_id;
        // 发布Request消息
        request_pub_->publish(request_msg);
    }

};

int main(int argc,char **argv)
{
  // 初始化ROS2客户端
  rclcpp::init(argc,argv);
  //调用spin函数,传入指针
  rclcpp::spin(std::make_shared<TwistBridge>());
  //释放资源
  rclcpp::shutdown();
  return 0;
}